lxd/auth/driver_openfga.go

```diff
@@ -0,0 +1,1079 @@
+	apiToken    string
+	storeID     string
+	authModelID string
+
+	client *client.OpenFgaClient
+}
+
+func (f *fga) configure(opts Opts) error {
+	if opts.config == nil {
+		return fmt.Errorf("Missing OpenFGA config")
+	}
+
+	val, ok := opts.config["openfga.api.token"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA API token")
+	}
+
+	f.apiToken = val.(string)
+
+	val, ok = opts.config["openfga.api.url"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA API URL")
+	}
+
+	f.apiURL = val.(string)
+
+	val, ok = opts.config["openfga.store.id"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA store ID")
+	}
+
+	f.storeID = val.(string)
+
+	val, ok = opts.config["openfga.store.model_id"]
+	if !ok {
+		return fmt.Errorf("Missing OpenFGA authorization model ID")
+	}
+
+	f.authModelID = val.(string)
+
+	if opts.resources == nil {
+		return fmt.Errorf("Missing resources for OpenFGA sync")
+	}
+
+	return nil
+}
+
+func (f *fga) load(ctx context.Context, opts Opts) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	err := f.configure(opts)
+	if err != nil {
+		return err
+	}
+
+	u, err := url.Parse(f.apiURL)
+	if err != nil {
+		return fmt.Errorf("Failed parsing URL: %w", err)
+	}
+
+	conf := client.ClientConfiguration{
+		ApiScheme: u.Scheme,
+		ApiHost:   u.Host,
+		StoreId:   f.storeID,
+		Credentials: &credentials.Credentials{
+			Method: credentials.CredentialsMethodApiToken,
+			Config: &credentials.Config{
+				ApiToken: f.apiToken,
+			},
+		},
+	}
+
+	f.client, err = client.NewSdkClient(&conf)
+	if err != nil {
+		return fmt.Errorf("Failed to create OpenFGA client: %w", err)
+	}
+
+	var builtinAuthorizationModel client.ClientWriteAuthorizationModelRequest
+
+	err = json.Unmarshal([]byte(authModel), &builtinAuthorizationModel)
+	if err != nil {
+		return err
+	}
+
+	// Compare existing model with built-in model
+	readModelResponse, err := f.client.ReadAuthorizationModel(ctx).Options(client.ClientReadAuthorizationModelOptions{
+		AuthorizationModelId: openfga.PtrString(f.authModelID),
+	}).Body(client.ClientReadAuthorizationModelRequest{}).Execute()
+	if err != nil {
+		return fmt.Errorf("Failed to read pre-existing OpenFGA model: %w", err)
+	} else if readModelResponse.AuthorizationModel == nil {
+		return fmt.Errorf("Pre-existing authorization model is missing")
+	}
+
+	if readModelResponse.AuthorizationModel.SchemaVersion != builtinAuthorizationModel.SchemaVersion {
+		return fmt.Errorf("Existing OpenFGA model has different schema version")
+	}
+
+	existingTypeDefinitions, err := json.Marshal(readModelResponse.AuthorizationModel.TypeDefinitions)
+	if err != nil {
+		return fmt.Errorf("Failed to compare OpenFGA model type definitions: %w", err)
+	}
+
+	builtinTypeDefinitions, err := json.Marshal(builtinAuthorizationModel.TypeDefinitions)
+	if err != nil {
+		return fmt.Errorf("Failed to compare OpenFGA model type definitions: %w", err)
+	}
+
+	if string(existingTypeDefinitions) != string(builtinTypeDefinitions) {
```

Comment:
Is there a risk of false negatives if the field order or whitespace varies but the logical content is the same ?

Reply:
The field order appears to be tied to the model. Both are of the same type and marshaled without indentation so whitespace should be fine. 

I did this because the logic for comparing the models was becoming quite cumbersome but you're right it isn't ideal. I will try again.